File: expensereport-nim/ExpenseReport.nim
type
  ExpenseType = enum DINNER, BREAKFAST, CAR_RENTAL
  Expense = tuple[expenseType: ExpenseType, amount: int]

proc printReport(expenses: seq[Expense]) =
  var meals: int = 0
  var total: int = 0
  echo "Expense Report"
  for i, expense in expenses:
    if expense.expenseType == DINNER or expense.expenseType == BREAKFAST:
      meals += expense.amount
    var name: string = ""
    case expense.expenseType
    of DINNER: name = "Dinner"
    of BREAKFAST: name = "Breakfast"
    of CAR_RENTAL: name = "Car Rental"
    let mealOverExpensesMarker = if expense.expenseType == DINNER and expense.amount > 5000 or expense.expenseType == BREAKFAST and expense.amount > 1000: "X" else: " "
    echo name, "\t", expense.amount, "\t", mealOverExpensesMarker
    total += expense.amount
  echo "Meal expenses: ", meals
  echo "Total expenses: ", total

printReport(@[(expenseType: DINNER, amount: 5000), (expenseType: DINNER, amount: 5001), (expenseType: BREAKFAST, amount: 1000), (expenseType: BREAKFAST, amount: 1001), (expenseType: CAR_RENTAL, amount: 4)])

File: expensereport-zig/ExpenseReport.zig
const std = @import("std");
const ExpenseType = enum { DINNER, BREAKFAST, CAR_RENTAL };
const print = std.debug.print;

const Expense = struct {
    type: ExpenseType,
    amount: u32,
};

fn printReport(expenses: []Expense) void {
    var total: u32 = 0;
    var meals: u32 = 0;

    print("Expense Report: {s}\n", .{""});

    for (expenses) |expense| {
        if (expense.type == ExpenseType.DINNER or expense.type == ExpenseType.BREAKFAST) {
            meals += expense.amount;
        }

        var name: []const u8 = "";
        switch (expense.type) {
            ExpenseType.DINNER => { name = "Dinner"; },
            ExpenseType.BREAKFAST => { name = "Breakfast"; },
            ExpenseType.CAR_RENTAL => { name = "Car Rental"; },
        }

        var mealOverExpensesMarker: []const u8 = if (expense.type == ExpenseType.DINNER and expense.amount > 5000 or expense.type == ExpenseType.BREAKFAST and expense.amount > 1000) "X" else " ";

        print("{s}\t{d}\t{s}\n", .{name, expense.amount, mealOverExpensesMarker});

        total += expense.amount;
    }

    print("Meal expenses: {d}\n", .{meals});
    print("Total expenses: {d}\n", .{total});
}

pub fn main() void {
    var expenses: [5]Expense = .{
        Expense {.type = ExpenseType.DINNER, .amount = 5000},
        Expense {.type = ExpenseType.DINNER, .amount = 5001},
        Expense {.type = ExpenseType.BREAKFAST, .amount = 1000},
        Expense {.type = ExpenseType.BREAKFAST, .amount = 1001},
        Expense {.type = ExpenseType.CAR_RENTAL, .amount = 4},
    };
    printReport(&expenses);
}
